### PR TITLE
perf: replace calldata slices by `abi.decode` to extract `ERC725X.execute` params

### PR DESCRIPTION
**Pull request by CJ42**
# What does this PR introduce?

## 🚀 Perfs

Replace calldata slices by `abi.decode` in the Key Manager for extracting the ERC725X execute parameters.

This reduce the deployment cost + runtime cost as follow (except for the main controller).

### Deployment cost

image

### Runtime cost

image

### PR Checklist

- [x] Wrote Tests
- [] Wrote & Generated Documentation (readme/natspec/dodoc)
- [x] Ran `npm run lint` && `npm run lint:solidity` (solhint)
- [x] Ran `npm run format` (prettier)
- [x] Ran `npm run build`
- [x] Ran `npm run test`

### CODE REVIEW

**Comment by Hugoo** [comment]
Task linked: [DEV-5007 Use `abi.decode` instead of manually deconstructing `ERC725X.execute()` calldatas in LSP6](https://app.clickup.com/t/2645698/DEV-5007)

File: contracts/LSP6KeyManager/LSP6KeyManagerCore.sol [hunk 1]
```diff
@@ -528,11 +528,34 @@ abstract contract LSP6KeyManagerCore is
...
+            // CHECK the offset of `data` is not pointing to the previous parameters
+            //
+            // offsets in calldata for ERC725X.execute(...) parameters (excluding function selector)
+            //
+            // - `operationType`: index 0 in calldata
+            // - `to`: index 32
+            // - `value`: index 64
+            // - `data`'s offset location: index 96
+            // - `data` starts at: index 128 (= 0x0000...0080)
+            if (bytes32(payload[100:132]) != bytes32(uint256(128))) {
```
Comment (b00ste) [inline contracts/LSP6KeyManager/LSP6KeyManagerCore.sol:540]: I am pretty sure that check is useless, we have tests for it, i suggest removing the check and run the test.

Comment (CJ42) [reply contracts/LSP6KeyManager/LSP6KeyManagerCore.sol:540]: @b00ste yes I believe so to it does not have the same impact anymore.
I will add some extra checks on when the offset of the `data` params is pointing backwards, to ensure the permissions check run depending on the controller permissions.

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 2]
```diff
@@ -219,39 +209,45 @@ abstract contract LSP6ExecuteModule {
 
         // CHECK if we are doing an empty call, as the receive() or fallback() function
         // of the controlledContract could run some code.
-        if (!hasSuperCall && !isCallDataPresent && !isTransferringValue) {
+        if (!hasSuperCall && isEmptyCall && !isTransferringValue) {
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:212]: Let's always put "isEmptyCall" as the first thing to check in the conditional to simplify the reading of the conditionals (and avoid context switching when moving from on conditional to another):

```suggestion
        if (isEmptyCall && !isTransferringValue && !hasSuperCall) {
```

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 3]
```diff
@@ -219,39 +209,45 @@ abstract contract LSP6ExecuteModule {
...
         // Skip if caller has SUPER permissions for external calls, with or without calldata (empty calls)
         if (hasSuperCall && !isTransferringValue) return;
 
         // Skip if caller has SUPER permission for value transfers
-        if (hasSuperTransferValue && !isCallDataPresent && isTransferringValue)
-            return;
+        if (hasSuperTransferValue && isEmptyCall && isTransferringValue) return;
 
         // Skip if both SUPER permissions are present
         if (hasSuperCall && hasSuperTransferValue) return;
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:227]: Move this as the first conditional check to reduce gas for main controller (if possible and tests pass?)

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 4]
```diff
@@ -219,39 +209,45 @@ abstract contract LSP6ExecuteModule {
...
             _requirePermissions(controller, permissions, _PERMISSION_CALL);
         }
 
         // Skip if caller has SUPER permissions for external calls, with or without calldata (empty calls)
         if (hasSuperCall && !isTransferringValue) return;
 
         // Skip if caller has SUPER permission for value transfers
-        if (hasSuperTransferValue && !isCallDataPresent && isTransferringValue)
-            return;
+        if (hasSuperTransferValue && isEmptyCall && isTransferringValue) return;
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:224]: Same here, emptyCall first, value Transfer 2nd, super permission check 3rd

```suggestion
        if (isEmptyCall && isTransferringValue  && hasSuperTransferValue) return;
```

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 5]
```diff
@@ -219,39 +209,45 @@ abstract contract LSP6ExecuteModule {
...
             _requirePermissions(controller, permissions, _PERMISSION_CALL);
         }
 
-        if (isCallDataPresent && !hasSuperCall) {
+        if (!isEmptyCall && !hasSuperCall) {
             _requirePermissions(controller, permissions, _PERMISSION_CALL);
         }
 
         // Skip if caller has SUPER permissions for external calls, with or without calldata (empty calls)
         if (hasSuperCall && !isTransferringValue) return;
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:221]: Same here, value check first, super call 2nd

if (!isTransferringValue && hasSuperCall) return;

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 6]
```diff
@@ -219,39 +209,45 @@ abstract contract LSP6ExecuteModule {
...
+        uint256 value,
+        bytes memory data
     ) internal view virtual {
-        (
-            uint256 operationType,
-            address to,
-            uint256 value,
-            bytes4 selector,
-            bool isEmptyCall
-        ) = _extractExecuteParameters(payload);
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:254]: We can now check and revert early, so move these two local variables defined after the error handling conditional below:

**Review by CJ42 — COMMENTED** [review]
Writing down some notes and ideas of refactoring of things to do tomorrow (to improve code readability + optimise for gas potentially)

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 7]
```diff
@@ -54,32 +54,17 @@ abstract contract LSP6ExecuteModule {
      * @param controlledContract the address of the ERC725 contract where the payload is executed and where the permissions are verified.
      * @param controller the address who want to run the execute function on the ERC725Account.
      * @param permissions the permissions of the controller address.
-     * @param payload the ABI encoded payload `controlledContract.execute(...)`.
      */
+    //  * @param payload the ABI encoded payload `controlledContract.execute(...)`.
```
Comment (b00ste) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:58]: I think you forgot to uncomment. Plus you might need to update `./docs`

File: contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol [hunk 8]
```diff
@@ -180,79 +170,79 @@ abstract contract LSP6ExecuteModule {
...
-        if (isTransferringValue && !hasSuperTransferValue) {
+        if (isValueTransfer && !hasSuperTransferValue) {
             _requirePermissions(
                 controller,
                 permissions,
                 _PERMISSION_TRANSFERVALUE
             );
         }
 
+        // Skip if both SUPER permissions are present
```
Comment (YamenMerhi) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6ExecuteModule.sol:209]: No big deal, but probably keep the `if .. return` after the checks to preserve how it was before the audit

**Review by YamenMerhi — APPROVED** [review]
looks good

**Review by skimaharvey — APPROVED** [review]
LGTM

**Pull request merged**